src/stcal/outlier_detection/utils.py

```diff
@@ -0,0 +1,337 @@
+    # the derivative image (dealing with interpolating issues?)
+    # and the standard n*sigma above the noise
+    threshold1 = scale1 * blot_deriv + snr1 * err_data
+    mask1 = np.greater(diff_noise, threshold1)
+
+    # Smooth the boolean mask with a 3x3 boxcar kernel
+    kernel = np.ones((3, 3), dtype=int)
+    mask1_smoothed = ndimage.convolve(mask1, kernel, mode='nearest')
+
+    # Create a 2nd boolean mask based on the 2nd set of
+    # scale and threshold values
+    threshold2 = scale2 * blot_deriv + snr2 * err_data
+    mask2 = np.greater(diff_noise, threshold2)
+
+    # Final boolean mask
+    return mask1_smoothed & mask2
+
+
+def gwcs_blot(median_data, median_wcs, blot_shape, blot_wcs, pix_ratio):
+    """
+    Resample the median data to recreate an input image based on
+    the blot wcs.
+
+    Parameters
+    ----------
+    median_data : numpy.ndarray
+        The data to blot.
+
+    median_wcs : gwcs.wcs.WCS
+        The wcs for the median data.
+
+    blot_shape : list of int
+        The target blot data shape.
+
+    blot_wcs : gwcs.wcs.WCS
+        The target/blotted wcs.
+
+    pix_ratio : float
+        Pixel ratio.
+
+    Returns
+    -------
+    blotted : numpy.ndarray
+        The blotted median data.
+
+    blot_img : datamodel
+        Datamodel containing header and WCS to define the 'blotted' image
+    """
+    # Compute the mapping between the input and output pixel coordinates
+    pixmap = calc_gwcs_pixmap(blot_wcs, median_wcs, blot_shape)
+    log.debug("Pixmap shape: {}".format(pixmap[:, :, 0].shape))
+    log.debug("Sci shape: {}".format(blot_shape))
+    log.info('Blotting {} <-- {}'.format(blot_shape, median_data.shape))
+
+    outsci = np.zeros(blot_shape, dtype=np.float32)
+
+    # Currently tblot cannot handle nans in the pixmap, so we need to give some
+    # other value.  -1 is not optimal and may have side effects.  But this is
+    # what we've been doing up until now, so more investigation is needed
+    # before a change is made.  Preferably, fix tblot in drizzle.
+    pixmap[np.isnan(pixmap)] = -1
+    tblot(median_data, pixmap, outsci, scale=pix_ratio, kscale=1.0,
+          interp='linear', exptime=1.0, misval=0.0, sinscl=1.0)
+
+    return outsci
+
+
+def calc_gwcs_pixmap(in_wcs, out_wcs, in_shape):
+    """
+    Return a pixel grid map from input frame to output frame.
+
+    Parameters
+    ----------
+    in_wcs : gwcs.wcs.WCS
+        Input/source wcs.
+
+    out_wcs : gwcs.wcs.WCS
+        Output/projected wcs.
+
+    in_shape : list of int
+        Input shape used to compute the input bounding box.
+
+    Returns
+    -------
+    pixmap : numpy.ndarray
+        Computed pixmap.
+    """
+    bb = wcs_bbox_from_shape(in_shape)
+    log.debug("Bounding box from data shape: {}".format(bb))
+
+    grid = gwcs.wcstools.grid_from_bounding_box(bb)
+    return np.dstack(reproject(in_wcs, out_wcs)(grid[0], grid[1]))
+
+
+def reproject(wcs1, wcs2):
```

Comment:
Similar to a comment you made to me, I wonder where this utility and similar ones, like `calc_gwcs_pixmap`, should go - tweakwcs? stcal alignment utils? here?  I'm fine merging this with it here, but perhaps a discussion we should open sooner rather than later.

Reply:
One more thing about this function I just realized when looking through open JWST PRs: it looks like the reproject function is getting either substantial changes or removal in the near future: https://github.com/spacetelescope/jwst/pull/8554

Reply:
Unfortunately the `calc_gwcs_pixmap` in ailgnment.utils uses a `reproject` function that isn't compatible with gwcs. Here's the line that failed when I tried to use it:
https://github.com/spacetelescope/stcal/blob/eb8528c21e0ffbbdec89f5e1d7323425850d3b69/src/stcal/alignment/util.py#L821
@nden mentioned that the function is using the FITS convention (which in this case fails for the gwcs.WCS).

Consolidating the many `reproject`s (and other duplicated functions) would certainly be an improvement. I think the `calc_gwcs_pixmap` in alignment.utils is unused (romancal has it's [own copy](https://github.com/spacetelescope/romancal/blob/459e8e345bbb1ba6bfd76d0d8a90f2058089791e/romancal/resample/resample_utils.py#L205)).

I guess this PR is a step in the wrong direction by adding another one but perhaps it's within scope of this PR to also delete the one in alignment utils or do you think that would make more sense in a separate PR?

Reply:
It would make sense to me to have a separate PR focused specifically on evaluating the use of all these utils here an in both jwst and romancal, and figure out if they should be moved to gwcs, consolidated and remain here, or removed entirely.  This could probably happen after this PR and my tweakreg PR are both merged, since both of them add more utils

Reply:
Thanks! I opened a issue for this: https://github.com/spacetelescope/stcal/issues/271